README.md
# duration
[![Build Status][1]][2]
[![Coverage Status][3]][4]
[![npm][5]][6]
[![dependencies Status][7]][8]
[![devDependencies Status][9]][10]
[![Downloads/week][11]][12]
[![Dependabot Status][13]][14]

[1]: https://travis-ci.org/gnarr/duration.svg?branch=master
[2]: https://travis-ci.org/gnarr/duration
[3]: https://coveralls.io/repos/github/gnarr/duration/badge.svg?branch=master
[4]: https://coveralls.io/github/gnarr/duration?branch=master
[5]: https://img.shields.io/npm/v/duration-converter.svg?style=flat-square
[6]: https://www.npmjs.com/package/duration-converter
[7]: https://david-dm.org/gnarr/duration.svg
[8]: https://david-dm.org/gnarr/duration
[9]: https://david-dm.org/gnarr/duration/dev-status.svg
[10]: https://david-dm.org/gnarr/duration?type=dev
[11]: https://img.shields.io/npm/dw/duration-converter.svg
[12]: https://www.npmjs.com/package/duration-converter
[13]: https://api.dependabot.com/badges/status?host=github&repo=gnarr/duration
[14]: https://dependabot.com

A small utility for converting between duration units. E.g '7 Weeks' to Micro Seconds.

## Installation

### Node.js

    $ npm install duration-converter

## Example usage:
```javascript
const { Duration } = require('duration-converter');

const sevenWeeks = new Duration('7 Weeks');
const milliseconds = sevenWeeks.MilliSeconds;

const threeDays = Duration.fromDays(3);
const hours = threeDays.Hours;

// will print out "4 days"
console.log(Duration.fromSeconds(345600).toString());

const a = new Date(2019, 3, 14);
const b = new Date(2019, 3, 15);
const betweenDates = Duration.between(a, b);

// will print 86400000
console.log(betweenDates.MilliSeconds);

```

## Warning!

Note that a year is always interpreted as being 365 days. There are no leap year calculations so going between units that cross the year "boundaries" will result in a small rounding error.


CHANGELOG.md
# Changelog

## 1.3.7
- Updated outdated dev dependencies.
## 1.3.6
- Dev dependency security updates.
## 1.3.5
- Updated outdated dev dependencies.
## 1.3.4
- Updated outdated dev dependencies.
- Added a test for 100% coverage.
## 1.3.3
- Republish.
## 1.3.2
- Added months as a supported duration unit. A month is defined as being one twelvth of a year.
## 1.3.1
- Fixed bugged npm release.
## 1.3.0
- Added between function that takes two dates and creates a duration between the dates.
- Updated dependencies.
## 1.2.9
- Added files property to package.json to only include compiled typescript and interface.
## 1.2.8
- Improved test coverage.
## 1.2.7
- add Changelog
## 1.2.6
- Update README.md.
- Added dependencies and devDependencies status badge.
## 1.2.5
- Added TS lint and Travis testing.
- Added test coverage reports.
- Added coverage badge to Readme.md.
- Improved test coverage for toString() and toSingular().
- Added Build Status badge.
- Added npm version and transpilation badges.
## 1.2.4
- Updated README.md.
## 1.2.3
- Fixed class exporting.
## 1.2.2
- Code ported to typescript.
- updated main and types paths.
## 1.2.1
- Updated README.
## 1.2.0
- Added from{Unit} for all time units.
## 1.1.0
- Added a static Duration.fromMilliseconds(input) function.
## 1.0.3
- Constructor now throws error with input in message.
## 1.0.2
- Added support for floating point numbers in constructor.
- Added expected failure tests
## 1.0.1
- Fixed floatingpoint formatting for human readable convertion.
- removed "use strict" from duration.js and added .jshintrc.
- Added tests.
## 1.0.0
- add ISC license


test.js
// Do this as the first thing so that any code reading it knows the right env.
process.env.NODE_ENV = 'test';

// Makes the script crash on unhandled rejections instead of silently
process.on('unhandledRejection', err => {
  throw err;
});

const jest = require('jest');
const argv = process.argv.slice(2);

// Watch unless on CI or in coverage mode
if (!process.env.CI && argv.indexOf('--coverage') < 0) {
  argv.push('--watch');
}

jest.run(argv);


src/duration.test.ts
import { Duration } from './duration';

test("convert nanoseconds to microseconds", () => {
  expect(new Duration("4000 nanoseconds").MicroSeconds).toBe(4);
});

test("convert single nanoseconds to microseconds", () => {
  expect(new Duration("1001 nanosecond").MicroSeconds).toBe(1.001);
});

test("convert microseconds to milliseconds", () => {
  expect(new Duration("9000 microseconds").MilliSeconds).toBe(9);
});

test("convert milliseconds to seconds", () => {
  expect(new Duration("6000 milliseconds").Seconds).toBe(6);
});

test("convert seconds to minutes", () => {
  expect(new Duration("120 seconds").Minutes).toBe(2);
});

test("convert minutes to hours", () => {
  expect(new Duration("360 minutes").Hours).toBe(6);
});

test("convert hours to days", () => {
  expect(new Duration("72 hours").Days).toBe(3);
});

test("convert days to weeks", () => {
  expect(new Duration("14 days").Weeks).toBe(2);
});

test("convert weeks to years", () => {
  expect(new Duration("365 days").Years).toBe(1);
});

test("convert days to years", () => {
  expect(new Duration("365 weeks").Years).toBe(7);
});

test("convert months to years", () => {
  expect(new Duration("36 months").Years).toBe(3);
});

test("convert a single month to days", () => {
  expect(new Duration("1 month").Days).toBe(365 / 12);
});

test("convert years to months", () => {
  expect(new Duration("1 year").Months).toBe(12);
});

test("convert years to decades", () => {
  expect(new Duration("30 years").Decades).toBe(3);
});

test("convert decades to centuries", () => {
  expect(new Duration("40 decades").Centuries).toBe(4);
});

test("convert centuries to millenniums", () => {
  expect(new Duration("50 centuries").Millenniums).toBe(5);
});

test("convert nanoseconds to millenniums", () => {
  expect(new Duration("63072000000000000000 nanoseconds").Millenniums).toBe(2);
});

test("convert millenniums to nanoseconds", () => {
  expect(new Duration("2 millenniums").NanoSeconds).toBe(63072000000000000000);
});

test("convert single millennium to nanoseconds", () => {
  expect(new Duration("1 millennium").NanoSeconds).toBe(31536000000000000000);
});

test("convert millenniums to centuries", () => {
  expect(new Duration("2 millenniums").Centuries).toBe(20);
});

test("convert centuries to decades", () => {
  expect(new Duration("6 centuries").Decades).toBe(60);
});

test("convert single century to decades", () => {
  expect(new Duration("1 century").Decades).toBe(10);
});

test("convert decades to years", () => {
  expect(new Duration("2 decades").Years).toBe(20);
});

test("convert single decade to years", () => {
  expect(new Duration("1 decade").Years).toBe(10);
});

test("convert years to weeks", () => {
  expect(new Duration("7 years").Weeks).toBe(365);
});

test("convert single year to weeks", () => {
  expect(new Duration("1 year").Weeks).toBe(52.142857142857146);
});

test("convert years to days", () => {
  expect(new Duration("5 years").Days).toBe(1825);
});

test("convert weeks to days", () => {
  expect(new Duration("5 weeks").Days).toBe(35);
});

test("convert single week to days", () => {
  expect(new Duration("1 week").Days).toBe(7);
});

test("convert days to hours", () => {
  expect(new Duration("2 days").Hours).toBe(48);
});

test("convert single day to hours", () => {
  expect(new Duration("1 day").Hours).toBe(24);
});

test("convert hours to minutes", () => {
  expect(new Duration("7 hours").Minutes).toBe(420);
});

test("convert single hour to minutes", () => {
  expect(new Duration("1 hour").Minutes).toBe(60);
});

test("convert minutes to seconds", () => {
  expect(new Duration("3 minutes").Seconds).toBe(180);
});

test("convert single minute to seconds", () => {
  expect(new Duration("1 minute").Seconds).toBe(60);
});

test("convert seconds to milliseconds", () => {
  expect(new Duration("7 seconds").MilliSeconds).toBe(7000);
});

test("convert single second to milliseconds", () => {
  expect(new Duration("1 second").MilliSeconds).toBe(1000);
});

test("convert milliseconds to microseconds", () => {
  expect(new Duration("2 milliseconds").MicroSeconds).toBe(2000);
});

test("convert single millisecond to microseconds", () => {
  expect(new Duration("1 millisecond").MicroSeconds).toBe(1000);
});

test("convert microseconds to nanoseconds", () => {
  expect(new Duration("5 microseconds").NanoSeconds).toBe(5000);
});

test("convert single microsecond to nanoseconds", () => {
  expect(new Duration("1 microsecond").NanoSeconds).toBe(1000);
});

test("convert floating point microseconds to nanoseconds", () => {
  expect(new Duration("8.432 microseconds").NanoSeconds).toBe(8432);
});

test("convert floating point microseconds to milliseconds", () => {
  expect(new Duration("0.432 microseconds").MilliSeconds).toBe(0.000432);
});

test("Convert to human readable form singular", () => {
  expect(new Duration("60 seconds").toString()).toBe("1 minute");
});

test("Convert to human readable form plural", () => {
  expect(new Duration("840 seconds").toString()).toBe("14 minutes");
});

test("Convert to human readable form zero", () => {
  expect(new Duration("0 microseconds").toString()).toBe("0 microseconds");
});

test("should fail on bad input", () => {
  expect(() => new Duration("20 meconds")).toThrow();
  expect(() => new Duration("10")).toThrow();
  expect(() => new Duration("minute")).toThrow();
});

test("should not fail on good input", () => {
  expect(() => new Duration("10488.469970703125 milliseconds")).not.toThrow();
});

test("should show input on error", () => {
  expect(() => new Duration("10488,469970703125 milliseconds")).toThrowError(
    "Error in input: '10488,469970703125 milliseconds'"
  );
});

test("should create a new instance from nanoseconds", () => {
  expect(Duration.fromNanoseconds(18000000).MilliSeconds).toBe(18);
});

test("should create a new instance from microseconds", () => {
  expect(Duration.fromMicroSeconds(18000).MilliSeconds).toBe(18);
});

test("should create a new instance from milliseconds", () => {
  expect(Duration.fromMilliseconds(18000).Seconds).toBe(18);
});

test("should create a new instance from seconds", () => {
  expect(Duration.fromSeconds(120).Minutes).toBe(2);
});

test("should create a new instance from minutes", () => {
  expect(Duration.fromMinutes(60).Seconds).toBe(3600);
});

test("should create a new instance from hours", () => {
  expect(Duration.fromHours(48).Days).toBe(2);
});

test("should create a new instance from days", () => {
  expect(Duration.fromDays(14).Weeks).toBe(2);
});

test("should create a new instance from weeks", () => {
  expect(Duration.fromWeeks(10).Days).toBe(70);
});

test("should create a new instance from months", () => {
  expect(Duration.fromMonths(120).Years).toBe(10);
});

test("should create a new instance from years", () => {
  expect(Duration.fromYears(20).Decades).toBe(2);
});

test("should create a new instance from decades", () => {
  expect(Duration.fromDecades(4).Years).toBe(40);
});

test("should create a new instance from centuries", () => {
  expect(Duration.fromCenturies(12).Years).toBe(1200);
});

test("should create a new instance from millenniums", () => {
  expect(Duration.fromMillenniums(2).Years).toBe(2000);
});

test("should give a single month when provided with months", () => {
// tslint:disable-next-line: no-string-literal
  expect(Duration["toSingular"]("Months")).toBe("Month");
});

test("should create a new instance from two dates", () => {
  const a = new Date(2019, 3, 14);
  const b = new Date(2019, 3, 15);
  expect(Duration.between(a, b).Hours).toBe(24);
});

test("should create a new instance from two timestamps", () => {
  const a = new Date(2019, 3, 14).getTime();
  const b = new Date(2019, 3, 15).getTime();
  expect(Duration.between(a, b).Hours).toBe(24);
});


test/duration.test.js
const Duration = require('../duration');

test('convert nanoseconds to microseconds', () => {
  expect(new Duration('4000 nanoseconds').MicroSeconds).toBe(4);
});

test('convert microseconds to milliseconds', () => {
  expect(new Duration('9000 microseconds').milliSeconds).toBe(9);
});

test('convert milliseconds to seconds', () => {
  expect(new Duration('6000 milliseconds').Seconds).toBe(6);
});

test('convert seconds to minutes', () => {
  expect(new Duration('120 seconds').Minutes).toBe(2);
});

test('convert minutes to hours', () => {
  expect(new Duration('360 minutes').Hours).toBe(6);
});

test('convert hours to days', () => {
  expect(new Duration('72 hours').Days).toBe(3);
});

test('convert days to weeks', () => {
  expect(new Duration('14 days').Weeks).toBe(2);
});

test('convert weeks to years', () => {
  expect(new Duration('365 days').Years).toBe(1);
});

test('convert days to years', () => {
  expect(new Duration('365 weeks').Years).toBe(7);
});

test('convert years to decades', () => {
  expect(new Duration('30 years').Decades).toBe(3);
});

test('convert decades to centuries', () => {
  expect(new Duration('40 decades').Centuries).toBe(4);
});

test('convert centuries to millenniums', () => {
  expect(new Duration('50 centuries').Millenniums).toBe(5);
});

test('convert nanoseconds to millenniums', () => {
  expect(new Duration('63072000000000000000 nanoseconds').Millenniums).toBe(2);
});

test('convert millenniums to nanoseconds', () => {
  expect(new Duration('2 millenniums').NanoSeconds).toBe(63072000000000000000);
});

test('convert millenniums to centuries', () => {
  expect(new Duration('2 millenniums').Centuries).toBe(20);
});

test('convert centuries to decades', () => {
  expect(new Duration('60 decades').Centuries).toBe(6);
});

test('convert decades to years', () => {
  expect(new Duration('20 years').Decades).toBe(2);
});

test('convert years to weeks', () => {
  expect(new Duration('7 years').Weeks).toBe(365);
});

test('convert years to days', () => {
  expect(new Duration('5 years').Days).toBe(1825);
});

test('convert weeks to days', () => {
  expect(new Duration('5 weeks').Days).toBe(35);
});

test('convert days to hours', () => {
  expect(new Duration('2 days').Hours).toBe(48);
});

test('convert hours to minutes', () => {
  expect(new Duration('7 hours').Minutes).toBe(420);
});

test('convert minutes to seconds', () => {
  expect(new Duration('3 minutes').Seconds).toBe(180);
});

test('convert seconds to milliseconds', () => {
  expect(new Duration('7 seconds').MilliSeconds).toBe(7000);
});

test('convert milliseconds to microseconds', () => {
  expect(new Duration('2 milliseconds').MicroSeconds).toBe(2000);
});

test('convert microseconds to nanoseconds', () => {
  expect(new Duration('5 microseconds').NanoSeconds).toBe(5000);
});

test('convert floating point microseconds to nanoseconds', () => {
  expect(new Duration('8.432 microseconds').NanoSeconds).toBe(8432);
});

test('convert floating point microseconds to milliseconds', () => {
  expect(new Duration('0.432 microseconds').MilliSeconds).toBe(0.000432);
});

test('Convert to human readable form', () => {
  expect(new Duration('840 seconds').toString()).toBe("14 minutes");
});

test('should fail on bad input', () => {
  expect(() => new Duration('20 meconds')).toThrow();
  expect(() => new Duration('10')).toThrow();
  expect(() => new Duration('minute')).toThrow();
  expect(() => new Duration(8000)).toThrow();
});

test('should not fail on good input', () => {
  expect(() => new Duration("10488.469970703125 milliseconds")).not.toThrow();
});

test('should show input on error', () => {
  expect(() => new Duration("10488,469970703125 milliseconds")).toThrowError("Error in input: '10488,469970703125 milliseconds'");
});

test('should create a new instance from nanoseconds', () => {
  expect(Duration.fromNanoseconds(18000000).MilliSeconds).toBe(18);
});

test('should create a new instance from microseconds', () => {
  expect(Duration.fromMicroSeconds(18000).MilliSeconds).toBe(18);
});

test('should create a new instance from milliseconds', () => {
  expect(Duration.fromMilliseconds(18000).Seconds).toBe(18);
});

test('should create a new instance from seconds', () => {
  expect(Duration.fromSeconds(120).Minutes).toBe(2);
});

test('should create a new instance from minutes', () => {
  expect(Duration.fromMinutes(60).Seconds).toBe(3600);
});

test('should create a new instance from hours', () => {
  expect(Duration.fromHours(48).Days).toBe(2);
});

test('should create a new instance from days', () => {
  expect(Duration.fromDays(14).Weeks).toBe(2);
});

test('should create a new instance from weeks', () => {
  expect(Duration.fromWeeks(10).Days).toBe(70);
});

test('should create a new instance from years', () => {
  expect(Duration.fromYears(20).Decades).toBe(2);
});

test('should create a new instance from decades', () => {
  expect(Duration.fromDecades(4).Years).toBe(40);
});

test('should create a new instance from centuries', () => {
  expect(Duration.fromCenturies(12).Years).toBe(1200);
});

test('should create a new instance from millenniums', () => {
  expect(Duration.fromMillenniums(2).Years).toBe(2000);
});



duration.js
function isSingular(value) {
  return value % 100 !== 11 && value % 10 === 1;
}
function toSingular(value) {
  return value.replace(/ies$/, 'y').replace(/s$/, '');
}
/**
 * Duration can easily convert between units. E.g '7 Weeks' to Micro Seconds.
 * Usage:
 * const sevenWeeks = new Duration('7 Weeks');
 * const milliseconds = sevenWeeks.MilliSeconds;
 *
 * @class Duration
 */
module.exports = class Duration {
  constructor(input) {
    if (!input) {
      this.milliSeconds = 0;
      return;
    }
    const parts = input.split(/([0-9]+\.?[0-9]*)/);
    const number = Number(parts[1]);
    if (isNaN(number) || parts[2].length === 0) {
      throw new Error(`Error in input: '${input}'`);
    }
    const unit = parts[2].trim().toLowerCase();
    switch(unit) {
      case 'nanosecond':
      case 'nanoseconds':
        this.NanoSeconds = number;
        return;
      case 'microsecond':
      case 'microseconds':
        this.MicroSeconds = number;
        return;
      case 'millisecond':
      case 'milliseconds':
        this.MilliSeconds = number;
        return;
      case 'second':
      case 'seconds':
        this.Seconds = number;
        return;
      case 'minute':
      case 'minutes':
        this.Minutes = number;
        return;
      case 'hour':
      case 'hours':
        this.Hours = number;
        return;
      case 'day':
      case 'days':
        this.Days = number;
        return;
      case 'week':
      case 'weeks':
        this.Weeks = number;
        return;
      case 'year':
      case 'years':
        this.Years = number;
        return;
      case 'decade':
      case 'decades':
        this.Decades = number;
        return;
      case 'century':
      case 'centuries':
        this.Centuries = number;
        return;
      case 'millennium':
      case 'millenniums':
      case 'millennia':
        this.Millenniums = number;
        return;
      default:
        throw new Error(`Error in input: '${input}'`);
    }
  }
  static fromNanoseconds(nanoseconds) {
    const duration = new Duration();
    duration.NanoSeconds = nanoseconds;
    return duration;
  }
  static fromMicroSeconds(microseconds) {
    const duration = new Duration();
    duration.MicroSeconds = microseconds;
    return duration;
  }
  static fromMilliseconds(milliSeconds) {
    const duration = new Duration();
    duration.MilliSeconds = milliSeconds;
    return duration;
  }
  static fromSeconds(seconds) {
    const duration = new Duration();
    duration.Seconds = seconds;
    return duration;
  }
  static fromMinutes(minutes) {
    const duration = new Duration();
    duration.Minutes = minutes;
    return duration;
  }
  static fromHours(hours) {
    const duration = new Duration();
    duration.Hours = hours;
    return duration;
  }
  static fromDays(days) {
    const duration = new Duration();
    duration.Days = days;
    return duration;
  }
  static fromWeeks(weeks) {
    const duration = new Duration();
    duration.Weeks = weeks;
    return duration;
  }
  static fromYears(years) {
    const duration = new Duration();
    duration.Years = years;
    return duration;
  }
  static fromDecades(decades) {
    const duration = new Duration();
    duration.Decades = decades;
    return duration;
  }
  static fromCenturies(centuries) {
    const duration = new Duration();
    duration.Centuries = centuries;
    return duration;
  }
  static fromMillenniums(millenniums) {
    const duration = new Duration();
    duration.Millenniums = millenniums;
    return duration;
  }
  get NanoSeconds() {
    return this.MicroSeconds * 1000;
  }
  set NanoSeconds(nanoSeconds) {
    this.MicroSeconds = nanoSeconds / 1000;
  }
  get MicroSeconds() {
    return this.milliSeconds * 1000;
  }
  set MicroSeconds(microSeconds) {
    this.milliSeconds = microSeconds / 1000;
  }
  get MilliSeconds() {
    return this.milliSeconds;
  }
  set MilliSeconds(milliSeconds) {
    this.milliSeconds = milliSeconds;
  }
  get Seconds() {
    return this.MilliSeconds / 1000;
  }
  set Seconds(seconds) {
    this.MilliSeconds = seconds * 1000;
  }
  get Minutes() {
    return this.Seconds / 60;
  }
  set Minutes(minutes) {
    this.Seconds = minutes * 60;
  }
  get Hours() {
    return this.Minutes / 60;
  }
  set Hours(hours) {
    this.Minutes = hours * 60;
  }
  get Days() {
    return this.Hours / 24;
  }
  set Days(days) {
    this.Hours = days * 24;
  }
  get Weeks() {
    return this.Days / 7;
  }
  set Weeks(weeks) {
    this.Days = weeks * 7;
  }
  get Years() {
    return this.Days / 365;
  }
  set Years(years) {
    this.Days = years * 365;
  }
  get Decades() {
    return this.Years / 10;
  }
  set Decades(decades) {
    this.Years = decades * 10;
  }
  get Centuries() {
    return this.Decades / 10;
  }
  set Centuries(centuries) {
    this.Decades = centuries * 10;
  }
  get Millenniums() {
    return this.Centuries / 10;
  }
  set Millenniums(millenniums) {
    this.Centuries = millenniums * 10;
  }
  static getters() {
    return ['NanoSeconds', 'MicroSeconds', 'MilliSeconds', 'Seconds', 'Minutes', 'Hours', 'Days', 'Weeks', 'Years', 'Decades', 'Centuries', 'Millenniums'];
  }
  toString() {
    for (const getter of Duration.getters().reverse()) {
      const test = this[getter];
      if (test >= 1 ) {
        const unit = isSingular(test) ? toSingular(getter) : getter;
        const formatted = test.toLocaleString(undefined, {maximumFractionDigits: 2});
        return `${formatted} ${unit.toLowerCase()}`;
      }
    }
    return `${this.MicroSeconds} microseconds`;
  }
};


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after the first failure
  // bail: false,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/tmp/jest_rs",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files usin a array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  moduleFileExtensions: [
    "ts",
    "tsx",
    "js",
    "json",
    "jsx",
    "node"
  ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "always",

  // A preset that is used as a base for Jest's configuration
  // preset: null,

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: null,

  // A list of paths to directories that Jest should use to search for files in
  roots: [
    "<rootDir>/src"
  ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // The path to a module that runs some code to configure or set up the testing framework before each test
  // setupTestFrameworkScriptFile: null,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.js?(x)",
  //   "**/?(*.)+(spec|test).js?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern Jest uses to detect test files
  testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.tsx?$",

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  transform: {
    "^.+\\.tsx?$": "ts-jest"
  },

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


src/duration.ts
export interface IDuration {
  NanoSeconds: number;
  MicroSeconds: number;
  MilliSeconds: number;
  Seconds: number;
  Minutes: number;
  Hours: number;
  Days: number;
  Weeks: number;
  Months: number;
  Years: number;
  Decades: number;
  Centuries: number;
  Millenniums: number;
}

/**
 * Duration can easily convert between units. E.g '7 Weeks' to Micro Seconds.
 * Usage:
 * const sevenWeeks = new Duration('7 Weeks');
 * const milliseconds = sevenWeeks.MilliSeconds;
 *
 * @class Duration
 */
export class Duration implements IDuration {
  get NanoSeconds() {
    return this.MicroSeconds * 1000;
  }
  set NanoSeconds(nanoSeconds) {
    this.MicroSeconds = nanoSeconds / 1000;
  }
  get MicroSeconds() {
    return this.milliSeconds * 1000;
  }
  set MicroSeconds(microSeconds) {
    this.milliSeconds = microSeconds / 1000;
  }
  get MilliSeconds() {
    return this.milliSeconds;
  }
  set MilliSeconds(milliSeconds) {
    this.milliSeconds = milliSeconds;
  }
  get Seconds() {
    return this.MilliSeconds / 1000;
  }
  set Seconds(seconds) {
    this.MilliSeconds = seconds * 1000;
  }
  get Minutes() {
    return this.Seconds / 60;
  }
  set Minutes(minutes) {
    this.Seconds = minutes * 60;
  }
  get Hours() {
    return this.Minutes / 60;
  }
  set Hours(hours) {
    this.Minutes = hours * 60;
  }
  get Days() {
    return this.Hours / 24;
  }
  set Days(days) {
    this.Hours = days * 24;
  }
  get Weeks() {
    return this.Days / 7;
  }
  set Weeks(weeks) {
    this.Days = weeks * 7;
  }
  get Months() {
    return this.Years * 12;
  }
  set Months(months) {
    this.Years = months / 12;
  }
  get Years() {
    return this.Days / 365;
  }
  set Years(years) {
    this.Days = years * 365;
  }
  get Decades() {
    return this.Years / 10;
  }
  set Decades(decades) {
    this.Years = decades * 10;
  }
  get Centuries() {
    return this.Decades / 10;
  }
  set Centuries(centuries) {
    this.Decades = centuries * 10;
  }
  get Millenniums() {
    return this.Centuries / 10;
  }
  set Millenniums(millenniums) {
    this.Centuries = millenniums * 10;
  }
  public static fromNanoseconds(nanoseconds: number) {
    const duration = new Duration();
    duration.NanoSeconds = nanoseconds;
    return duration;
  }
  public static fromMicroSeconds(microseconds: number) {
    const duration = new Duration();
    duration.MicroSeconds = microseconds;
    return duration;
  }
  public static fromMilliseconds(milliSeconds: number) {
    const duration = new Duration();
    duration.MilliSeconds = milliSeconds;
    return duration;
  }
  public static fromSeconds(seconds: number) {
    const duration = new Duration();
    duration.Seconds = seconds;
    return duration;
  }
  public static fromMinutes(minutes: number) {
    const duration = new Duration();
    duration.Minutes = minutes;
    return duration;
  }
  public static fromHours(hours: number) {
    const duration = new Duration();
    duration.Hours = hours;
    return duration;
  }
  public static fromDays(days: number) {
    const duration = new Duration();
    duration.Days = days;
    return duration;
  }
  public static fromWeeks(weeks: number) {
    const duration = new Duration();
    duration.Weeks = weeks;
    return duration;
  }
  public static fromMonths(months: number) {
    const duration = new Duration();
    duration.Months = months;
    return duration;
  }
  public static fromYears(years: number) {
    const duration = new Duration();
    duration.Years = years;
    return duration;
  }
  public static fromDecades(decades: number) {
    const duration = new Duration();
    duration.Decades = decades;
    return duration;
  }
  public static fromCenturies(centuries: number) {
    const duration = new Duration();
    duration.Centuries = centuries;
    return duration;
  }
  public static fromMillenniums(millenniums: number) {
    const duration = new Duration();
    duration.Millenniums = millenniums;
    return duration;
  }
  public static between(a: Date | number, b: Date | number) {
    const duration = new Duration();
    const aMilliseconds = a instanceof Date ? a.getTime() : a;
    const bMilliseconds = b instanceof Date ? b.getTime() : b;
    duration.MilliSeconds = Math.abs(aMilliseconds - bMilliseconds);
    return duration;
  }
  public static getters() {
    return [
      "NanoSeconds",
      "MicroSeconds",
      "MilliSeconds",
      "Seconds",
      "Minutes",
      "Hours",
      "Days",
      "Weeks",
      "Months",
      "Years",
      "Decades",
      "Centuries",
      "Millenniums",
    ];
  }

  private static isSingular(value: number) {
    return value % 100 !== 11 && value % 10 === 1;
  }
  private static toSingular(value: string) {
    return value.replace(/ies$/, "y").replace(/s$/, "");
  }
  private milliSeconds: number;
  public constructor(input?: string) {
    this.milliSeconds = 0;
    if (!input) {
      return;
    }
    const parts = input.split(/([0-9]+\.?[0-9]*)/);
    const n = Number(parts[1]);
    if (isNaN(n) || parts[2].length === 0) {
      throw new Error(`Error in input: '${input}'`);
    }
    const unit = parts[2].trim().toLowerCase();
    switch (unit) {
      case "nanosecond":
      case "nanoseconds":
        this.NanoSeconds = n;
        return;
      case "microsecond":
      case "microseconds":
        this.MicroSeconds = n;
        return;
      case "millisecond":
      case "milliseconds":
        this.MilliSeconds = n;
        return;
      case "second":
      case "seconds":
        this.Seconds = n;
        return;
      case "minute":
      case "minutes":
        this.Minutes = n;
        return;
      case "hour":
      case "hours":
        this.Hours = n;
        return;
      case "day":
      case "days":
        this.Days = n;
        return;
      case "week":
      case "weeks":
        this.Weeks = n;
        return;
      case "month":
      case "months":
        this.Months = n;
        return;
      case "year":
      case "years":
        this.Years = n;
        return;
      case "decade":
      case "decades":
        this.Decades = n;
        return;
      case "century":
      case "centuries":
        this.Centuries = n;
        return;
      case "millennium":
      case "millenniums":
      case "millennia":
        this.Millenniums = n;
        return;
      default:
        throw new Error(`Error in input: '${input}'`);
    }
  }

  public toString() {
    for (const getter of Duration.getters().reverse()) {
      const test = this[getter as (keyof IDuration)];
      if (test >= 1) {
        const unit = Duration.isSingular(test)
          ? Duration.toSingular(getter)
          : getter;
        const formatted = test.toLocaleString(undefined, {
          maximumFractionDigits: 2,
        });
        return `${formatted} ${unit.toLowerCase()}`;
      }
    }
    return `${this.MicroSeconds} microseconds`;
  }
}
